### PR TITLE
A simple BF interpreter.

### PR DESCRIPTION
**Pull request by ltratt**
The only slightly clever thing this does is to precalculate the jumps for '[' and ']' (and, unlike some other BF implementations I've seen, store it as a Vec rather than a HashMap).

Note that, despite it's name, this is not yet Yoricked. I thought it would be useful to just get a basic BF interpreter up and running before I worried about anything Yorick-specific.

If you want to try some examples out, have a look at e.g. https://github.com/kostya/benchmarks/tree/master/brainfuck2 or http://www.hevanet.com/cristofd/brainfuck/. For example:

```
$ https://raw.githubusercontent.com/kostya/benchmarks/master/brainfuck2/bench.b
$ cargo run --release bench.b
    Finished release [optimized] target(s) in 0.01s                           
     Running `target/release/ykbf bench.b`
ZYXWVUTSRQPONMLKJIHGFEDCBA
```

### CODE REVIEW

File: Cargo.toml
Comment (vext01) [inline Cargo.toml:6]: Probably should add a license directive.

File: src/main.rs
Comment (vext01) [inline src/main.rs]: Are there not Rust-native alternatives for get/putchar?

File: src/main.rs
Comment (vext01) [inline src/main.rs:23]: If I understand, all of this is just to get the basename of argv[0]?

File: src/main.rs
Comment (vext01) [inline src/main.rs]: I don't understand the allocation logic here. You make enough space for possibly every position, then truncate it to zero?

File: src/main.rs
Comment (vext01) [inline src/main.rs:63]: Can we add comments to these vars.

`pc` is obvious, but I'm not sure what `ptr` or `store` are.

File: src/main.rs
Comment (vext01) [inline src/main.rs:133]: A nitpick, but what does `inv` stand for here? It's the input bytes, right?

File: src/main.rs
Comment (vext01) [inline src/main.rs:118]: Is the double increment in the case of `>` intentional?

**Review by vext01 — COMMENTED** [review]
Some shallow comments.

In the next bout, I'll have read about bf a bit.

File: Cargo.toml
Comment (ltratt) [reply Cargo.toml:6]: Agreed.

File: src/main.rs
Comment (ltratt) [reply src/main.rs]: I think I'd have to convert things to a UTF-8 string, where BF only deals in 8 bit bytes. I'm happy to be proven wrong though!

File: src/main.rs
Comment (ltratt) [reply src/main.rs:23]: Yes.

File: src/main.rs
Comment (ltratt) [reply src/main.rs]: Huh? `with_capacity` mallocs memory but doesn't fill it. `resize` (on the next line) then fills it.

File: src/main.rs
Comment (ltratt) [reply src/main.rs:63]: I could call them `ptr` and `tape` perhaps? I must admit, I think they're fairly obvious if you know a Turing tape / BF (and if you don't know BF, would comments help?).

File: src/main.rs
Comment (ltratt) [reply src/main.rs:133]: What would you call it? It's a minor temporary variable: its name doesn't really seem *that* important to me.

File: src/main.rs
Comment (ltratt) [reply src/main.rs:118]: Yes.

File: src/main.rs
Comment (vext01) [reply src/main.rs]: Ah, my bad, I misread.

How about:
```
let map = vec![0; out.len()];
```

File: src/main.rs
Comment (vext01) [reply src/main.rs:63]: Let's place this on hold until I've read about bf.

File: src/main.rs
Comment (vext01) [reply src/main.rs]: A quick google doesn't turn anything up, so I can live with the FFI calls.

File: src/main.rs
Comment (vext01) [reply src/main.rs:133]: It's not that important at all. I'd have called it `bytes` perhaps.

File: src/main.rs
Comment (vext01) [reply src/main.rs:118]: Ah, they are different pointers in fact. Disregard my comment.

File: src/main.rs
Comment (ltratt) [reply src/main.rs]: Good idea.

File: src/main.rs
Comment (vext01) [reply src/main.rs:63]: I've skimmed the wikipedia page for bf and it uses mixed terminology. I don't feel strongly to be honest, but if you did change your mind, maybe `cells`?

Comment (vext01) [reply src/main.rs:63]: (instead of store)

**Comment by vext01** [comment]
Might be nice to have a dumb test which runs some simple program. Just so that we can verify the build without fetching an external input.

But I won't insist, since it's only bf :P Up to you.

File: src/main.rs
Comment (ltratt) [reply src/main.rs:63]: OK.

**Comment by vext01** [comment]
Any news on this one?

**Comment by ltratt** [comment]
No.

**Comment by ltratt** [comment]
OK, this is ready for another review. I think I've addressed your comments. I've also added test and run travis on PRs.

File: src/main.rs
Comment (vext01) [inline src/main.rs]: Good catch.

I assume `EOF` is `-1`? Should we use `libc::EOF`?

https://docs.rs/libc/0.2.46/libc/constant.EOF.html

Errors are ignored, should we panic?

**Comment by vext01** [comment]
Just one comment.

File: src/main.rs
Comment (ltratt) [reply src/main.rs]: Agreed on `EOF`. I'm not sure about errors. Do you have an opinion?

Comment (ltratt) [reply src/main.rs]: @vext01 ping

Comment (vext01) [reply src/main.rs]: Why not panic? Otherwise the output is incorrect.

Comment (ltratt) [reply src/main.rs]: OK, the new commit does that. It's more verbose than I expected!

File: src/main.rs
Comment (vext01) [inline src/main.rs]: How about using `fdopen` to convert `STDIN_FILENO` to a file, instead of opening stderr a second time?

> The fdopen() function associates a stream with the existing file descriptor fildes.

File: src/main.rs
Comment (vext01) [inline src/main.rs]: What's this line for? It makes a `CString`, but doesn't assign it?

Comment (ltratt) [reply src/main.rs]: No, it actually makes it possible to `free` it.

File: src/main.rs
Comment (ltratt) [reply src/main.rs]: Huh?

Comment (vext01) [reply src/main.rs]: I'm an idiot. Disregard this thread.

File: src/main.rs
Comment (vext01) [reply src/main.rs]: Ohhhh. I see.

Could we use `mode.as_ptr()` instead of `mode.into_raw()`, then Rust retains ownership, and `from_raw` shouldn't be necessary?

(might be wrong)

Comment (ltratt) [reply src/main.rs]: Yes, we can! See the new commit. Good spot.

Comment (vext01) [reply src/main.rs]: Beautiful!

**Comment by vext01** [comment]
Time to squash

**Review by vext01 — APPROVED** [review]

**Comment by ltratt** [comment]
Squashed.

**Comment by vext01** [comment]
Is it me, or are tests running twice?

**Comment by ltratt** [comment]
There's the branch and PR tests, and then GH adds extra buttons on top of Travis. I don't know why. It's the same on most of our repos.

**Comment by vext01** [comment]
Is it because you have travis enabled twice? Once as a webhook, once as an integration?

**Comment by ltratt** [comment]
Have a look in the settings and you'll see why. It doesn't do any harm, and it's been like this on most of our repos for 6+ months.

**Comment by vext01** [comment]
> Have a look in the settings and you'll see why.

Well, yes, that's how I figured it out.

Why don't you disable the web-hook? I don't see any need to run the tests twice.

**Comment by vext01** [comment]
Wait, it'd be the service you'd want to kill:

> GitHub Services are being deprecated.

**Comment by ltratt** [comment]
If you want to sort this out, be my guest. I'm fine with things as they are, and I don't think it should hold this PR up, unless you want to experiment with things on your own. If so, be my guest, but I'll leave it to you ;)

**Pull request merged**
